### Top 8 Powerful Functions in Python for deeper Data Analysis

1. Groupby
2. pivot_table
3. Aggregate
4. cross_tab
5. concat
6. apply
7. lambda
8. del (Keyword)

In [3]:
import pandas as pd
insurance_data=pd.read_csv("Insurance.csv")
insurance_data

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520
...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830
1334,18,female,31.920,0,no,northeast,2205.98080
1335,18,female,36.850,0,no,southeast,1629.83350
1336,21,female,25.800,0,no,southwest,2007.94500


### 1.1 Perform initial analysis

### Checking how many numerical and object data types are there

In [ ]:
# Checking how many rows and columns are there

In [6]:
insurance_data.shape

(1338, 7)

In [155]:
insurance_data.dtypes

age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
EC           object
dtype: object

When client gave the data there were 1338 records , and 7 features (This is very important)

### What was the minimum value for the numerical data in the database?

**Descriptive statistics**


In [4]:
insurance_data.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


Minimum age is 18 and max is 64
Minimum bmi is 15.96 and max is 53.13
Minimum childre is 0 and max is 5
Minimum charges are 1121.87 and max is 63370

In [5]:
insurance_data.describe(include="all")

,age,sex,bmi,children,smoker,region,charges
count,1338.000000,1338,1338.000000,1338.000000,1338,1338,1338.000000
unique,NaN,2,NaN,NaN,2,4,NaN
top,NaN,male,NaN,NaN,no,southeast,NaN
freq,NaN,676,NaN,NaN,1064,364,NaN
mean,39.207025,NaN,30.663397,1.094918,NaN,NaN,13270.422265
std,14.049960,NaN,6.098187,1.205493,NaN,NaN,12110.011237
min,18.000000,NaN,15.960000,0.000000,NaN,NaN,1121.873900
25%,27.000000,NaN,26.296250,0.000000,NaN,NaN,4740.287150
50%,39.000000,NaN,30.400000,1.000000,NaN,NaN,9382.033000
75%,51.000000,NaN,34.693750,2.000000,NaN,NaN,16639.912515


**Checking the features of the tables**

In [8]:
insurance_data.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')

**Checking the total null values in each column**

In [9]:
insurance_data.isna().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [ ]:
#there are no null values based on the above analysis

**Checking duplicate values**

In [16]:
insurance_data[insurance_data.duplicated()]

,age,sex,bmi,children,smoker,region,charges
581,19,male,30.59,0,no,northwest,1639.5631


**Finding the tow duplicate value in which all rows**

In [19]:
insurance_data[(insurance_data["age"]==19) & (insurance_data["charges"]==1639.5631)]

,age,sex,bmi,children,smoker,region,charges
195,19,male,30.59,0,no,northwest,1639.5631
581,19,male,30.59,0,no,northwest,1639.5631


#pd.set_option("display.max_columns",None)  ----- dont use this coz it configure notebook to always display max columns which takes time to load

### Pull up the insurance charges based on region

### Tell me the count of unique regions? Overall how many category is there?

In [34]:
insurance_data["region"].nunique()

4

### Tell me the different unique regions? Not count?

In [35]:
insurance_data["region"].unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

### Pull up the average insurance charges based on region?

In [44]:
insurance_data.groupby("region")["charges"].mean().round(2).to_frame()

,charges
region,
northeast,13406.38
northwest,12417.58
southeast,14735.41
southwest,12346.94


Based on this southeast shows highest average insurance charges

### Pull up the average insurance charges based on region with ascending and descending order?

In [47]:
insurance_data.groupby("region")["charges"].mean().round(2).sort_values()

region
southwest    12346.94
northwest    12417.58
northeast    13406.38
southeast    14735.41
Name: charges, dtype: float64

In [49]:
insurance_data.groupby("region")["charges"].mean().round(2).sort_values(ascending=False)

region
southeast    14735.41
northeast    13406.38
northwest    12417.58
southwest    12346.94
Name: charges, dtype: float64

### How man insurance have been taken in ech of these regions?

In [43]:
insurance_data.groupby("region")["region"].count()

region
northeast    324
northwest    325
southeast    364
southwest    325
Name: region, dtype: int64

this shows that south east taken more insurance than other region
while the other two regions had taken less insurance

### Pull up the count based on region and sex?

In [53]:
insurance_data.groupby(["region","sex"])["region"].count()

region     sex   
northeast  female    161
           male      163
northwest  female    164
           male      161
southeast  female    175
           male      189
southwest  female    162
           male      163
Name: region, dtype: int64

### Discrete and continous data?

Discrete data: countable/can be segmented in to disstinct values
Continous data: measurable, may have untis and cannot be segmented

Continuous data cannot be naturally segmented because it flows smoothly without gaps. 
To segment it, you have to artificially force it into ranges or bins (like grouping ages into "20–29" and "30–39")

### Pull up the average insurance charges based on region and sex?

In [60]:
insurance_data.groupby(["region","sex"])["charges"].mean().round(2)

region     sex   
northeast  female    12953.20
           male      13854.01
northwest  female    12479.87
           male      12354.12
southeast  female    13499.67
           male      15879.62
southwest  female    11274.41
           male      13412.88
Name: charges, dtype: float64

### In group by waht type of data needs to be taken and what data gives error? And outside groupby what kind of data should be taken to perform a aggregate operation?

groupby(["region","sex"])  -- You have to take only the discrete data
insurance_data.groupby(["region","sex"])["charges"].mean().round(2) - and outside group data when you have to perform aggregate operation you need to take only the continous data. Here we have taken charges column to perform the mean operation it cant be done on discrete data.

### Pull up the max insurance charges paid based on region and sex?

In [63]:
insurance_data.groupby(["region","sex"])["charges"].max().round(2).sort_values(ascending=False)

region     sex   
southeast  female    63770.43
           male      62592.87
northwest  male      60021.40
northeast  female    58571.07
northwest  female    55135.40
southwest  male      52590.83
           female    48824.45
northeast  male      48549.18
Name: charges, dtype: float64

### Pivot table

### Please pivot the average insurance charges paid based on region and sex?

In [70]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index="region",
    columns="sex")

sex,female,male
region,,
northeast,12953.203151,13854.005374
northwest,12479.870397,12354.119575
southeast,13499.669243,15879.617173
southwest,11274.411264,13412.883576


Here the index will be row and the things which needs to be categorized can be given here
Index and column can be a discrete features
Values that should come under rows and infex
default aggfunction is mean so you get the mean above
you can also change the sort for the above

In breif you need two categorical items to perform a aggregate function on a continous variable while pivoting

### Please pivot the count of insurance charges paid based on region as row and sex as column? Also please find max insurance charges with the same method?

In [71]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index="region",
    columns="sex",
                aggfunc="count")

sex,female,male
region,,
northeast,161,163
northwest,164,161
southeast,175,189
southwest,162,163


In [72]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index="region",
    columns="sex",
                aggfunc="max")

sex,female,male
region,,
northeast,58571.07448,48549.17835
northwest,55135.40209,60021.39897
southeast,63770.42801,62592.87309
southwest,48824.45000,52590.82939


In [73]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index="region",
    columns="sex",
                aggfunc="min")

sex,female,male
region,,
northeast,2196.47320,1694.7964
northwest,2117.33885,1621.3402
southeast,1607.51010,1121.8739
southwest,1727.78500,1241.5650


### Please pivot the average insurance charges paid based on region category as row and sex+children as column? 

In [74]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index="region",
    columns=["sex","children"])

sex              female                                            \
children              0             1             2             3   
region                                                              
northeast   9789.822330  17124.689856  14523.222213  13730.098756   
northwest  12583.872074  11118.855757  13127.592952  13430.708082   
southeast  14862.013008   9869.908240  11766.313671  19722.982900   
southwest  10149.745507  10841.643945  16230.264372   8438.736294   

sex                                           male                \
children              4            5             0             1   
region                                                             
northeast  14542.416663          NaN  13389.637372  15474.289174   
northwest  11024.418200   8965.79575   9812.969532   9515.042120   
southeast  18267.100687   9923.88820  13791.806750  17424.652082   
southwest  10945.677243  10023.64775  13727.264465   9971.325962   

sex                                                                
children              2             3             4             5  
region                                                             
northeast  12317.910591  14935.224532  14442.275463   6978.973483  
northwest  13762.553939  22971.223279  11508.318988           NaN  
southeast  18647.954693  17101.818726   8726.908900  10306.994883  
southwest  18875.953545  12071.592329  17923.948000   6864.669500

In [76]:
pd.pivot_table(data=insurance_data,
    values="charges",
    index=["region","children"],
    columns=["sex"])

sex                       female          male
region    children                            
northeast 0          9789.822330  13389.637372
          1         17124.689856  15474.289174
          2         14523.222213  12317.910591
          3         13730.098756  14935.224532
          4         14542.416663  14442.275463
          5                  NaN   6978.973483
northwest 0         12583.872074   9812.969532
          1         11118.855757   9515.042120
          2         13127.592952  13762.553939
          3         13430.708082  22971.223279
          4         11024.418200  11508.318988
          5          8965.795750           NaN
southeast 0         14862.013008  13791.806750
          1          9869.908240  17424.652082
          2         11766.313671  18647.954693
          3         19722.982900  17101.818726
          4         18267.100687   8726.908900
          5          9923.888200  10306.994883
southwest 0         10149.745507  13727.264465
          1         10841.643945   9971.325962
          2         16230.264372  18875.953545
          3          8438.736294  12071.592329
          4         10945.677243  17923.948000
          5         10023.647750   6864.669500

### Aggregate function

In [77]:
insurance_data.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


### you need to get sum, mean, min and maximum values for insurance charges grouped by region and sex?

In [91]:
insurance_data.groupby(["region","sex"])["charges"].aggregate(func=["count","mean","max","min","sum"]).round(2)

count      mean       max      min         sum
region    sex                                                   
northeast female    161  12953.20  58571.07  2196.47  2085465.71
          male      163  13854.01  48549.18  1694.80  2258202.88
northwest female    164  12479.87  55135.40  2117.34  2046698.75
          male      161  12354.12  60021.40  1621.34  1989013.25
southeast female    175  13499.67  63770.43  1607.51  2362442.12
          male      189  15879.62  62592.87  1121.87  3001247.65
southwest female    162  11274.41  48824.45  1727.78  1826454.62
          male      163  13412.88  52590.83  1241.56  2186300.02

In [94]:
insurance_data.groupby(["region","sex"])["charges"].agg(func=["count","mean","max","min","sum"]).round(2)

count      mean       max      min         sum
region    sex                                                   
northeast female    161  12953.20  58571.07  2196.47  2085465.71
          male      163  13854.01  48549.18  1694.80  2258202.88
northwest female    164  12479.87  55135.40  2117.34  2046698.75
          male      161  12354.12  60021.40  1621.34  1989013.25
southeast female    175  13499.67  63770.43  1607.51  2362442.12
          male      189  15879.62  62592.87  1121.87  3001247.65
southwest female    162  11274.41  48824.45  1727.78  1826454.62
          male      163  13412.88  52590.83  1241.56  2186300.02

### you need to get sum, mean, min and maximum values for insurance charges grouped by region and sex? You need to give the name for the functions that is called not the default name?

In [95]:
insurance_data.groupby(["region","sex"])["charges"].agg(func=[("Count","count"),("Average","mean"),("Maximum value","max"),("Minimum value","min"),("Total","sum")]).round(2)

Count   Average  Maximum value  Minimum value       Total
region    sex                                                              
northeast female    161  12953.20       58571.07        2196.47  2085465.71
          male      163  13854.01       48549.18        1694.80  2258202.88
northwest female    164  12479.87       55135.40        2117.34  2046698.75
          male      161  12354.12       60021.40        1621.34  1989013.25
southeast female    175  13499.67       63770.43        1607.51  2362442.12
          male      189  15879.62       62592.87        1121.87  3001247.65
southwest female    162  11274.41       48824.45        1727.78  1826454.62
          male      163  13412.88       52590.83        1241.56  2186300.02

### 4. Crosstab

In [96]:
insurance_data.head()


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


### How many smokers and non-smokers based on region?

In [101]:
insurance_data.groupby(["region","smoker"]).count() #this shows wrong data so as all are discrete columns you need somethin else

age  sex  bmi  children  charges
region    smoker                                  
northeast no      257  257  257       257      257
          yes      67   67   67        67       67
northwest no      267  267  267       267      267
          yes      58   58   58        58       58
southeast no      273  273  273       273      273
          yes      91   91   91        91       91
southwest no      267  267  267       267      267
          yes      58   58   58        58       58

In [104]:
pd.crosstab(index=insurance_data["region"], columns=insurance_data["smoker"])

smoker,no,yes
region,,
northeast,257,67
northwest,267,58
southeast,273,91
southwest,267,58


### I need total smokers for each region along with subtotals of total people for the each region?

In [105]:
pd.crosstab(index=insurance_data["region"], columns=insurance_data["smoker"],margins=True)

smoker,no,yes,All
region,,,
northeast,257,67,324
northwest,267,58,325
southeast,273,91,364
southwest,267,58,325
All,1064,274,1338


### Why we are not using data=insurance data?

we dont have that argument to pass hence via index we are passing the selected row or column from the ttable

### What is the main use of crosstab how it differs from pivot and groupby?

Groupby and pivot table need a continous variable to perform a aggregagate operation and provide result. 
It can have many categories but it needs a continous data to give proper result

Crosstab can compare the two categorical features to returns their frequencies.

this is a function of panda and not a function of dataframe/Series

### Can we use aggregate function for the two categorical data?

In [ ]:
 #no you cant use as it shows the below error so you have to always take a continous variable to perform the aggregate funct

In [108]:
pd.crosstab(index=insurance_data["region"], columns=insurance_data["smoker"],margins=True,aggfunc="mean")

ValueError: aggfunc cannot be used without values.

In [114]:
pd.crosstab(index=insurance_data["region"], columns=insurance_data["smoker"],values=insurance_data["charges"],margins=True,aggfunc="mean").round()

smoker,no,yes,All
region,,,
northeast,9166.0,29674.0,13406.0
northwest,8556.0,30192.0,12418.0
southeast,8032.0,34845.0,14735.0
southwest,8019.0,32269.0,12347.0
All,8434.0,32050.0,13270.0


### Can you take more than 1 category in index or columns?

No you cant take

In [115]:
pd.crosstab(index=insurance_data[["region","sex"]], columns=insurance_data["smoker"],values=insurance_data["charges"],margins=True,aggfunc="mean").round()

ValueError: 2

### Concat

In [119]:
sales_data_2017=pd.read_csv("Sales-Transactions-2017.csv.csv")
sales_data_2018=pd.read_csv("Sales-Transactions-2018.csv.csv")
sales_data_2019=pd.read_csv("Sales-Transactions-2019.csv.csv")
sales_data_edited=pd.read_csv("Sales-Transactions-Edited.csv.csv") #concatinated and datacleaned data

### How to concatenate all three 201,2018 and 2019 csv file and what are the two points to be remembered

1. All the csv which are going to be concatenated should have same columns if you are appending by rows
2. it will append the rows to the requested table
   
   

In [122]:
Sales_full_data=pd.concat(objs=[sales_data_2017,sales_data_2018,sales_data_2019])

In [125]:
Sales_full_data.shape

(111206, 9)

In [126]:
sales_data_edited.shape

(95562, 6)

### Merge vs Concat?

1. Concat appends data row wise - this extend/expand the table horizontally
2. Merge appends data based on columns - this extend/expand the table vertically

### insurance_data["age"] any alternative for this?

In [ ]:
#can be used only for one column but the above can be used for multiple columns, for cross tab as it takes one category only this can be used

In [127]:
insurance_data.age 

0       19
1       18
2       28
3       33
4       32
        ..
1333    50
1334    18
1335    18
1336    21
1337    61
Name: age, Length: 1338, dtype: int64

### Apply

In [128]:
insurance_data.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


### 6.1 create a new column with a name "Eligibility criteria" if the age > =20 = eligible else not eligible?

### Can i try if else condition to get this?

it cant take the data from a data frmae and print the result so its not feasible, 
then we need to use one more function called as apply

In [130]:
if insurance_data["age"]>20:
    print("Eliagible")
else:
    print("Not eligible")

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

### How to use apply to apply a column? is apply a panda function or a data structure function? What we have to add in apply functuin?

Apply is a DataFrame function
We need to add user defined function as a argument

In [137]:
def get(x):
        if x>20:
            print("Eliagible")
        else:
            print("Not eligible")

In [138]:

insurance_data["age"].apply(get)

Not eligible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Not eligible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Not eligible
Not eligible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Eliagible
Elia

0       None
1       None
2       None
3       None
4       None
        ... 
1333    None
1334    None
1335    None
1336    None
1337    None
Name: age, Length: 1338, dtype: object

### How to assign the above to a new column?

In [1]:
insurance_data["Eligibilty criteria"]=insurance_data["age"].apply(get)


NameError: name 'insurance_data' is not defined

In [140]:
insurance_data

,age,sex,bmi,children,smoker,region,charges,Eligibilty criteria
0,19,female,27.900,0,yes,southwest,16884.92400,None
1,18,male,33.770,1,no,southeast,1725.55230,None
2,28,male,33.000,3,no,southeast,4449.46200,None
3,33,male,22.705,0,no,northwest,21984.47061,None
4,32,male,28.880,0,no,northwest,3866.85520,None
...,...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830,None
1334,18,female,31.920,0,no,northeast,2205.98080,None
1335,18,female,36.850,0,no,southeast,1629.83350,None
1336,21,female,25.800,0,no,southwest,2007.94500,None


### Why it is giving the value as none for new column?

Because in get user defined function we used print instead of return 
Return is very important to change the original data

In [141]:
def get(x):
        if x>20:
            return "Eliagible"
        else:
            return "Not eligible"

In [142]:
insurance_data["Eligibility criteria"]=insurance_data.age.apply(get)

In [143]:
insurance_data

,age,sex,bmi,children,smoker,region,charges,Eligibilty criteria,Eligibility criteria
0,19,female,27.900,0,yes,southwest,16884.92400,None,Not eligible
1,18,male,33.770,1,no,southeast,1725.55230,None,Not eligible
2,28,male,33.000,3,no,southeast,4449.46200,None,Eliagible
3,33,male,22.705,0,no,northwest,21984.47061,None,Eliagible
4,32,male,28.880,0,no,northwest,3866.85520,None,Eliagible
...,...,...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830,None,Eliagible
1334,18,female,31.920,0,no,northeast,2205.98080,None,Not eligible
1335,18,female,36.850,0,no,southeast,1629.83350,None,Not eligible
1336,21,female,25.800,0,no,southwest,2007.94500,None,Eliagible


In [144]:
insurance_data.Eligibility=insurance_data.age.apply(get)

C:\Users\Q848957\AppData\Local\Temp\ipykernel_37268\2940740992.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  insurance_data.Eligibility=insurance_data.age.apply(get)


### When is apply function used?

WHen you want to apply a condition for the values of the dataframe and store in a new column apply can be used.

### Lambda function

### What is lambda function

1. Anonymous/nameless function (no need to give any name for the function as we give for user defined)
2. Single line expression should be given for the function

### Example for the lambda function for the below line how can we replace the user defined fucntion with the lambda?
insurance_data["Eligibility criteria"]=insurance_data.age.apply(get)

In [147]:
insurance_data["EC"]=insurance_data.age.apply(lambda x:"Eligible" if x>20 else "Not eligible")
insurance_data

,age,sex,bmi,children,smoker,region,charges,Eligibilty criteria,Eligibility criteria,EC
0,19,female,27.900,0,yes,southwest,16884.92400,None,Not eligible,Not eligible
1,18,male,33.770,1,no,southeast,1725.55230,None,Not eligible,Not eligible
2,28,male,33.000,3,no,southeast,4449.46200,None,Eliagible,Eligible
3,33,male,22.705,0,no,northwest,21984.47061,None,Eliagible,Eligible
4,32,male,28.880,0,no,northwest,3866.85520,None,Eliagible,Eligible
...,...,...,...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830,None,Eliagible,Eligible
1334,18,female,31.920,0,no,northeast,2205.98080,None,Not eligible,Not eligible
1335,18,female,36.850,0,no,southeast,1629.83350,None,Not eligible,Not eligible
1336,21,female,25.800,0,no,southwest,2007.94500,None,Eliagible,Eligible


### How to create a function in x variable where two numbers are given and it compares and says greater or lesser? using lambda function?

In [148]:
x=lambda a,b:"Greater" if a>b else "Lesser"
print(x(4,5))

Lesser


### What is list comprehension?

List comprehension is a concise, single-line syntax used to create new lists from existing iterables like lists, tuples, or ranges

### Del

### What is DEl is it a keyword or function

Del is a keyword not a function

### How to delete columns Eligibility criteria	and EC in insurance data?

In [153]:
del insurance_data["Eligibilty criteria"]

In [154]:
insurance_data

,age,sex,bmi,children,smoker,region,charges,EC
0,19,female,27.900,0,yes,southwest,16884.92400,Not eligible
1,18,male,33.770,1,no,southeast,1725.55230,Not eligible
2,28,male,33.000,3,no,southeast,4449.46200,Eligible
3,33,male,22.705,0,no,northwest,21984.47061,Eligible
4,32,male,28.880,0,no,northwest,3866.85520,Eligible
...,...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830,Eligible
1334,18,female,31.920,0,no,northeast,2205.98080,Not eligible
1335,18,female,36.850,0,no,southeast,1629.83350,Not eligible
1336,21,female,25.800,0,no,southwest,2007.94500,Eligible


### Can u delete multiple columns using del?

No you cannot delete multiple columns